### ISSUE TITLE
Running mutationCoverage fails with Pitest 1.4.5

### ISSUE THREAD

**Issue by surli** [STAMP-project/pitest-descartes#95]
#### Characteristics
- **Issue Type**: bug
- **Reproducibility**: always
- **Severity**: crash
- **Tool/Service/Component**: Descartes
- **Execution Environment**: Linux

#### Description

We cannot execute Descartes with Pitest after upgrading it to 1.4.5. We are using Descartes 1.2.4 but I also checked with Descartes 1.2.5-SNAPSHOT with the same error:

> Caused by: java.lang.NoSuchMethodError: org.pitest.mutationtest.config.TestPluginArguments.<init>(Ljava/lang/String;Lorg/pitest/testapi/TestGroupConfig;Ljava/util/Collection;Ljava/util/Collection;Z)V

See here the full stack trace:
<details>
[ERROR] Failed to execute goal org.pitest:pitest-maven:1.4.5:mutationCoverage (pitest-check) on project xwiki-commons-xml: Execution pitest-check of goal org.pitest:pitest-maven:1.4.5:mutationCoverage failed: An API incompatibility was encountered while executing org.pitest:pitest-maven:1.4.5:mutationCoverage: java.lang.NoSuchMethodError: org.pitest.mutationtest.config.TestPluginArguments.<init>(Ljava/lang/String;Lorg/pitest/testapi/TestGroupConfig;Ljava/util/Collection;Ljava/util/Collection;Z)V
[ERROR] 
[ERROR] -> [Help 1]
org.apache.maven.lifecycle.LifecycleExecutionException: Failed to execute goal org.pitest:pitest-maven:1.4.5:mutationCoverage (pitest-check) on project xwiki-commons-xml: Execution pitest-check of goal org.pitest:pitest-maven:1.4.5:mutationCoverage failed: An API incompatibility was encountered while executing org.pitest:pitest-maven:1.4.5:mutationCoverage: java.lang.NoSuchMethodError: org.pitest.mutationtest.config.TestPluginArguments.<init>(Ljava/lang/String;Lorg/pitest/testapi/TestGroupConfig;Ljava/util/Collection;Ljava/util/Collection;Z)V
-----------------------------------------------------
realm =    plugin>org.pitest:pitest-maven:1.4.5
strategy = org.codehaus.plexus.classworlds.strategy.SelfFirstStrategy
-----------------------------------------------------

    at org.apache.maven.lifecycle.internal.MojoExecutor.execute (MojoExecutor.java:215)
    at org.apache.maven.lifecycle.internal.MojoExecutor.execute (MojoExecutor.java:156)
    at org.apache.maven.lifecycle.internal.MojoExecutor.execute (MojoExecutor.java:148)
    at org.apache.maven.lifecycle.internal.LifecycleModuleBuilder.buildProject (LifecycleModuleBuilder.java:117)
    at org.apache.maven.lifecycle.internal.LifecycleModuleBuilder.buildProject (LifecycleModuleBuilder.java:81)
    at org.apache.maven.lifecycle.internal.builder.singlethreaded.SingleThreadedBuilder.build (SingleThreadedBuilder.java:56)
    at org.apache.maven.lifecycle.internal.LifecycleStarter.execute (LifecycleStarter.java:128)
    at org.apache.maven.DefaultMaven.doExecute (DefaultMaven.java:305)
    at org.apache.maven.DefaultMaven.doExecute (DefaultMaven.java:192)
    at org.apache.maven.DefaultMaven.execute (DefaultMaven.java:105)
    at org.apache.maven.cli.MavenCli.execute (MavenCli.java:956)
    at org.apache.maven.cli.MavenCli.doMain (MavenCli.java:288)
    at org.apache.maven.cli.MavenCli.main (MavenCli.java:192)
    at sun.reflect.NativeMethodAccessorImpl.invoke0 (Native Method)
    at sun.reflect.NativeMethodAccessorImpl.invoke (NativeMethodAccessorImpl.java:62)
    at sun.reflect.DelegatingMethodAccessorImpl.invoke (DelegatingMethodAccessorImpl.java:43)
    at java.lang.reflect.Method.invoke (Method.java:498)
    at org.codehaus.plexus.classworlds.launcher.Launcher.launchEnhanced (Launcher.java:289)
    at org.codehaus.plexus.classworlds.launcher.Launcher.launch (Launcher.java:229)
    at org.codehaus.plexus.classworlds.launcher.Launcher.mainWithExitCode (Launcher.java:415)
    at org.codehaus.plexus.classworlds.launcher.Launcher.main (Launcher.java:356)
Caused by: org.apache.maven.plugin.PluginExecutionException: Execution pitest-check of goal org.pitest:pitest-maven:1.4.5:mutationCoverage failed: An API incompatibility was encountered while executing org.pitest:pitest-maven:1.4.5:mutationCoverage: java.lang.NoSuchMethodError: org.pitest.mutationtest.config.TestPluginArguments.<init>(Ljava/lang/String;Lorg/pitest/testapi/TestGroupConfig;Ljava/util/Collection;Ljava/util/Collection;Z)V
-----------------------------------------------------
realm =    plugin>org.pitest:pitest-maven:1.4.5
strategy = org.codehaus.plexus.classworlds.strategy.SelfFirstStrategy
-----------------------------------------------------

    at org.apache.maven.plugin.DefaultBuildPluginManager.executeMojo (DefaultBuildPluginManager.java:186)
    at org.apache.maven.lifecycle.internal.MojoExecutor.execute (MojoExecutor.java:210)
    at org.apache.maven.lifecycle.internal.MojoExecutor.execute (MojoExecutor.java:156)
    at org.apache.maven.lifecycle.internal.MojoExecutor.execute (MojoExecutor.java:148)
    at org.apache.maven.lifecycle.internal.LifecycleModuleBuilder.buildProject (LifecycleModuleBuilder.java:117)
    at org.apache.maven.lifecycle.internal.LifecycleModuleBuilder.buildProject (LifecycleModuleBuilder.java:81)
    at org.apache.maven.lifecycle.internal.builder.singlethreaded.SingleThreadedBuilder.build (SingleThreadedBuilder.java:56)
    at org.apache.maven.lifecycle.internal.LifecycleStarter.execute (LifecycleStarter.java:128)
    at org.apache.maven.DefaultMaven.doExecute (DefaultMaven.java:305)
    at org.apache.maven.DefaultMaven.doExecute (DefaultMaven.java:192)
    at org.apache.maven.DefaultMaven.execute (DefaultMaven.java:105)
    at org.apache.maven.cli.MavenCli.execute (MavenCli.java:956)
    at org.apache.maven.cli.MavenCli.doMain (MavenCli.java:288)
    at org.apache.maven.cli.MavenCli.main (MavenCli.java:192)
    at sun.reflect.NativeMethodAccessorImpl.invoke0 (Native Method)
    at sun.reflect.NativeMethodAccessorImpl.invoke (NativeMethodAccessorImpl.java:62)
    at sun.reflect.DelegatingMethodAccessorImpl.invoke (DelegatingMethodAccessorImpl.java:43)
    at java.lang.reflect.Method.invoke (Method.java:498)
    at org.codehaus.plexus.classworlds.launcher.Launcher.launchEnhanced (Launcher.java:289)
    at org.codehaus.plexus.classworlds.launcher.Launcher.launch (Launcher.java:229)
    at org.codehaus.plexus.classworlds.launcher.Launcher.mainWithExitCode (Launcher.java:415)
    at org.codehaus.plexus.classworlds.launcher.Launcher.main (Launcher.java:356)
Caused by: org.apache.maven.plugin.PluginContainerException: An API incompatibility was encountered while executing org.pitest:pitest-maven:1.4.5:mutationCoverage: java.lang.NoSuchMethodError: org.pitest.mutationtest.config.TestPluginArguments.<init>(Ljava/lang/String;Lorg/pitest/testapi/TestGroupConfig;Ljava/util/Collection;Ljava/util/Collection;Z)V

    at org.apache.maven.plugin.DefaultBuildPluginManager.executeMojo (DefaultBuildPluginManager.java:184)
    at org.apache.maven.lifecycle.internal.MojoExecutor.execute (MojoExecutor.java:210)
    at org.apache.maven.lifecycle.internal.MojoExecutor.execute (MojoExecutor.java:156)
    at org.apache.maven.lifecycle.internal.MojoExecutor.execute (MojoExecutor.java:148)
    at org.apache.maven.lifecycle.internal.LifecycleModuleBuilder.buildProject (LifecycleModuleBuilder.java:117)
    at org.apache.maven.lifecycle.internal.LifecycleModuleBuilder.buildProject (LifecycleModuleBuilder.java:81)
    at org.apache.maven.lifecycle.internal.builder.singlethreaded.SingleThreadedBuilder.build (SingleThreadedBuilder.java:56)
    at org.apache.maven.lifecycle.internal.LifecycleStarter.execute (LifecycleStarter.java:128)
    at org.apache.maven.DefaultMaven.doExecute (DefaultMaven.java:305)
    at org.apache.maven.DefaultMaven.doExecute (DefaultMaven.java:192)
    at org.apache.maven.DefaultMaven.execute (DefaultMaven.java:105)
    at org.apache.maven.cli.MavenCli.execute (MavenCli.java:956)
    at org.apache.maven.cli.MavenCli.doMain (MavenCli.java:288)
    at org.apache.maven.cli.MavenCli.main (MavenCli.java:192)
    at sun.reflect.NativeMethodAccessorImpl.invoke0 (Native Method)
    at sun.reflect.NativeMethodAccessorImpl.invoke (NativeMethodAccessorImpl.java:62)
    at sun.reflect.DelegatingMethodAccessorImpl.invoke (DelegatingMethodAccessorImpl.java:43)
    at java.lang.reflect.Method.invoke (Method.java:498)
    at org.codehaus.plexus.classworlds.launcher.Launcher.launchEnhanced (Launcher.java:289)
    at org.codehaus.plexus.classworlds.launcher.Launcher.launch (Launcher.java:229)
    at org.codehaus.plexus.classworlds.launcher.Launcher.mainWithExitCode (Launcher.java:415)
    at org.codehaus.plexus.classworlds.launcher.Launcher.main (Launcher.java:356)
Caused by: java.lang.NoSuchMethodError: org.pitest.mutationtest.config.TestPluginArguments.<init>(Ljava/lang/String;Lorg/pitest/testapi/TestGroupConfig;Ljava/util/Collection;Ljava/util/Collection;Z)V
    at org.pitest.mutationtest.config.ReportOptions.createMinionSettings (ReportOptions.java:601)
    at org.pitest.mutationtest.config.SettingsFactory.createCoverageOptions (SettingsFactory.java:115)
    at org.pitest.mutationtest.tooling.EntryPoint.execute (EntryPoint.java:97)
    at org.pitest.mutationtest.tooling.EntryPoint.execute (EntryPoint.java:50)
    at org.pitest.maven.RunPitStrategy.execute (RunPitStrategy.java:35)
    at org.pitest.maven.AbstractPitMojo.analyse (AbstractPitMojo.java:463)
    at org.pitest.maven.AbstractPitMojo.execute (AbstractPitMojo.java:404)
    at org.apache.maven.plugin.DefaultBuildPluginManager.executeMojo (DefaultBuildPluginManager.java:137)
    at org.apache.maven.lifecycle.internal.MojoExecutor.execute (MojoExecutor.java:210)
    at org.apache.maven.lifecycle.internal.MojoExecutor.execute (MojoExecutor.java:156)
    at org.apache.maven.lifecycle.internal.MojoExecutor.execute (MojoExecutor.java:148)
    at org.apache.maven.lifecycle.internal.LifecycleModuleBuilder.buildProject (LifecycleModuleBuilder.java:117)
    at org.apache.maven.lifecycle.internal.LifecycleModuleBuilder.buildProject (LifecycleModuleBuilder.java:81)
    at org.apache.maven.lifecycle.internal.builder.singlethreaded.SingleThreadedBuilder.build (SingleThreadedBuilder.java:56)
    at org.apache.maven.lifecycle.internal.LifecycleStarter.execute (LifecycleStarter.java:128)
    at org.apache.maven.DefaultMaven.doExecute (DefaultMaven.java:305)
    at org.apache.maven.DefaultMaven.doExecute (DefaultMaven.java:192)
    at org.apache.maven.DefaultMaven.execute (DefaultMaven.java:105)
    at org.apache.maven.cli.MavenCli.execute (MavenCli.java:956)
    at org.apache.maven.cli.MavenCli.doMain (MavenCli.java:288)
    at org.apache.maven.cli.MavenCli.main (MavenCli.java:192)
    at sun.reflect.NativeMethodAccessorImpl.invoke0 (Native Method)
    at sun.reflect.NativeMethodAccessorImpl.invoke (NativeMethodAccessorImpl.java:62)
    at sun.reflect.DelegatingMethodAccessorImpl.invoke (DelegatingMethodAccessorImpl.java:43)
    at java.lang.reflect.Method.invoke (Method.java:498)
    at org.codehaus.plexus.classworlds.launcher.Launcher.launchEnhanced (Launcher.java:289)
    at org.codehaus.plexus.classworlds.launcher.Launcher.launch (Launcher.java:229)
    at org.codehaus.plexus.classworlds.launcher.Launcher.mainWithExitCode (Launcher.java:415)
    at org.codehaus.plexus.classworlds.launcher.Launcher.main (Launcher.java:356)
</details>

**Comment by oscarlvp** [comment]
Thanks for reporting :)
At this moment we are not supporting PIT 1.4.5. The dependency is set to PIT 1.4.5 in version 1.2.4 and the 1.2.5-SNAPSHOT uses PIT 1.4.3.
We will upgrade for the next release.

**Comment by surli** [comment]
Hmm from what I see there has been a release for descartes few weeks ago. Shouldn't this one be closed as fixed?

**Comment by oscarlvp** [comment]
@surli if the issue is not happening anymore on your side then we can close it.

**Comment by surli** [comment]
> @surli if the issue is not happening anymore on your side then we can close it.

ok, I'll try with latest version then, and close it if needed.

**Comment by surli** [comment]
Apparently it's indeed working well now. Thanks!

**Comment by surli** [comment]
FTR we obtained different mutation scores when upgrading pit-descartes from 1.2.4 to 1.2.5 and PIT from 1.4.0 to 1.4.7.

**Comment by vmassol** [comment]
Yeah this is breaking our build ATM since our threshold are wrong now without us changing the code. @oscarlvp any idea where this is coming from and whether it's normal or not. Thx!

**Comment by surli** [comment]
So in order to understand what happened I tried to run PIT&descartes with various version, to see where the mutation score changed.

I don't have much info but I've discovered this:
  - with descartes 1.2.4 I've been able to run PIT 1.4.0 -> 1.4.2 without any change in the mutation score. Starting with PIT 1.4.3 there's apparently a change in the API so I'm not able to run it with descartes 1.2.4
  - with descartes 1.2.5 I've been able to run PIT 1.4.4 -> 1.4.7 with always the changed mutation score. 

Note that apparently PIT 1.4.3 is not supported by any version of descartes. So my guess is that the mutation score changed with the descartes following change (extracted from RN):
> The following method patterns have been added to the stop methods:
>   * return this
>   * methods that only return a parameter: public int m(int x, int y) { return y; }
>    simple setters returning this: public A setX(int x){ this.x = x; return this; }

Now I might be wrong and it could be related to PIT changes, hard to say. WDYT @oscarlvp ?
